### ISSUE TITLE
Calculation Error

### ISSUE THREAD

**Issue by angangi** [GPUPeople/spECK#6]
I found that when calculating webbase-1M matrix on 1080Ti, turning on cuSparse check will reports a calculation error, the same happens for pkustk12 matrix, the printed info is similar to the following:

Row: 321772 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321773 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321774 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321775 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321776 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321777 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321778 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321779 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321780 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321781 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321782 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321783 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/169

**Comment by dabeschte** [comment]
Hi!

what version of CUDA and which branch of this repository are you using? In CUDA 11, I was not able to multiply pkustk12 with cusparse, because of resource requirements. Even if cusparse were able to finish, the memory on the 1080Ti is too small to keep both input and output matrices on the device on the same time when using double precision. Have you tried float instead of double? That would probably fit into memory.

But I was able to process webbase-1M and the results matched. The resulting matrices are also much smaller - memory should not be a problem here. 
Can you give me more details? Do other matrices work correctly?

**Comment by angangi** [comment]
Sorry for not giving you the detail information. I'm using Ubuntu 14.04.6 LTS, CUDA10.0 (which I guess is the error reason) and the master branch, and got many calculation errors. More error test cases are listed below, grateful if you can do some tests on those matrices:

scircuit
webbase-1M
pkustk12
c8_mat11
cont1_l (which takes too long to get the result)
dbir2
dblp-2010
exdata_1
language
net150
vsp_finan512_scagr7-2c_rlfddd

**Comment by dabeschte** [comment]
The product of cont1_l is pretty large and does not fit into memory. 
However, the rest of the matrices work on my system (Windows 11, RTX 3090 24GB, CUDA 11.3) and they did also work back when I did my evaluations for the paper (Ubuntu 18.04, CUDA 10.1, Titan V 12GB). 

I am really suprised that you get incorrect results / freezes rather than exceptions.
It is a bit difficult to debug this remotely since it is running correctly on my system... But did you try compiling it in Debug mode instead of Release? Does that make a difference / improve error messages?
If you find any more clues about the source of the error, that would be really helpful.

Besides that, I can only recommend trying the code on a more recent setup (OS and CUDA) if you have one available somewhere.

**Comment by angangi** [comment]
OK，thanks for your recommendation！And  I'd to reconfirm that you have turn the `CompareResult ` to true in the .ini file? If so，I would like to close the issue and continue to debug by myself to find the problem, Thanks again!

**Comment by dabeschte** [comment]
Yes, CompareResult is set to true.

Btw. I have one more question: could you try to also set TrackIndividualTimes=true and post the result for one of the matrices that finished, but showed incorrect results?

Eventually, this could show whether of the kernels is skipped/crashed completely.

Sorry for not being able to help you out here. I hope that you still find a solution, because I am pretty confident that the project is quite stable now...but of course there could still be some bugs that only happen in rare cases or on older CUDA version.

**Comment by angangi** [comment]
For example:
webbase-1M.mtx

Matrix: 1000005x1000005: 3105536 nonzeros
spECK     initial mallocs = 6.328384 ms
spECK  count computations = 1.499488 ms
spECK       load-balancer = 1.923680 ms
spECK      GlobalMaps Cnt = 0.385984 ms
spECK     counting kernel = 4.969952 ms
spECK        malloc mat C = 0.505312 ms
spECK   num load-balancer = 0.208384 ms
spECK     init GlobalMaps = 0.003520 ms
spECK      numeric kernel = 51.654305 ms
spECK      Sorting kernel = 14.060864 ms
spECK             cleanup = 0.938112 ms

spECK            complete = 82.532349 ms

Row: 321019 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321020 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321021 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321022 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321023 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321120 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321121 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321122 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321123 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321124 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321125 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321126 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321127 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321128 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696
Row: 321129 | Row indices do NOT match: (Ref|Comp) : (31841|0) - pos: 0/1696

**Comment by dabeschte** [comment]
I managed to get a 1080 Ti and I am able to reproduce the error. Just wanted to give you a heads-up that I am looking for the error!

**Comment by dabeschte** [comment]
Fixed it.
Turned out to a missing __launch_bounds__ that I removed last time I fixed another bug, because they seemed unnecessary on my system. But they turn out to be required to fit all requested resources on Pascal devices. I would really appreciate if Nvidia would throw an exception or set an error flag at least when a kernel cannot be launched. What it actually does, is to silently skip execution...

I hope that these matrices now work on your system. I quickly checked on mine, and it seems to be fine. Only problem is that thw new version of cuSparse that came with cuda11 requires more resources and therefore is not able to process some of the matrices that spECK is able to process (but that also means, that we can't use cuSparse to check for correctness). I have not checked my code with cuda10 yet (but I did update the branch). Let me know if everything is working now :)
